sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3957,3 +3957,194 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
+            if (array1.isNullAt(i) && !foundNullElement) {
+              foundNullElement = true
+            } else {
+              val elem = array1.get(i, elementType)
+              hs.add(elem)
+            }
+            i += 1
+          }
+          i = 0
+          while (i < array2.numElements() && result) {
+            if (array2.isNullAt(i)) {
+              if (!foundNullElement) {
+                result = false
+              }
+            } else {
+              val elem = array2.get(i, elementType)
+              if (!hs.contains(elem)) {
+                result = false
+              }
+            }
+            i += 1
+          }
+          result
+        }
+    } else {
+      (array1, array2) =>
+        if (array2.numElements() == 0) {
+          true
+        } else if (array1.numElements() == 0) {
+          false
+        } else {
+          var alreadySeenNull = false
+          var i = 0
+          var elementFound = true
+          while (elementFound && i < array2.numElements()) {
+            var found = false
+            val elem2 = array2.get(i, elementType)
+            if (array2.isNullAt(i)) {
+              if (!alreadySeenNull) {
+                var j = 0
+                while (!found && j < array1.numElements()) {
+                  found = array1.isNullAt(j)
+                  j += 1
+                }
+                // array1 is scanned only once for null element
+                alreadySeenNull = true
+              }
+            } else {
+              var j = 0
+              while (!found && j < array2.numElements()) {
+                if (!array1.isNullAt(j)) {
+                  val elem1 = array1.get(j, elementType)
+                  if (ordering.equiv(elem2, elem1)) {
+                    found = true
+                  }
+                }
+                j += 1
+              }
+            }
+            if (!found) {
+              elementFound = false
+            }
+            i += 1
+          }
+          elementFound
+        }
+    }
+  }
+
+  override def nullSafeEval(input1: Any, input2: Any): Any = {
+    val array1 = input1.asInstanceOf[ArrayData]
+    val array2 = input2.asInstanceOf[ArrayData]
+
+    evalContains(array1, array2)
+  }
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val i = ctx.freshName("i")
+    val value = ctx.freshName("value")
+    if (canUseSpecializedHashSet) {
+      val jt = CodeGenerator.javaType(elementType)
+
+      nullSafeCodeGen(ctx, ev, (array1, array2) => {
+        val result = ctx.freshName("result")
+        val foundNullElement = ctx.freshName("foundNullElement")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$hsTypeName()"
+        val hashSet = ctx.freshName("hashSet")
+
+        def withArray1NullCheck(body: String): String =
+          s"""
+             |if ($array1.isNullAt($i) && !$foundNullElement) {
```

Comment:
If there is more than one null element in `array1`, does this code work?

Reply:
> If there is more than one null element in `array1`, does this code work?

Update UT and it works. also change t as below like  https://github.com/apache/spark/pull/30243#discussion_r528061382
```
             |if ($array1.isNullAt($i)) {
             |  if (!$foundNullElement) {
             |    $foundNullElement = true;
             |  }
```